core-backend/wasmi/src/funcs/internal.rs

```diff
@@ -16,34 +16,80 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
+use core::ops::{Deref, DerefMut};
+
+use codec::MaxEncodedLen;
+use gear_backend_common::memory::{
+    MemoryAccessManager, WasmMemoryRead, WasmMemoryReadAs, WasmMemoryReadDecoded, WasmMemoryWrite,
+    WasmMemoryWriteAs,
+};
+
 use super::*;
 use crate::state::State;
 
-pub(crate) struct CallerWrap<'a, E>(Caller<'a, HostState<E>>)
+pub(crate) struct CallerWrap<'a, E>
 where
     E: Ext + IntoExtInfo<E::Error> + 'static,
-    E::Error: IntoExtError;
+    E::Error: IntoExtError,
+{
+    caller: Caller<'a, HostState<E>>,
+    manager: MemoryAccessManager<E>,
+    memory: WasmiMemory,
+}
 
-impl<'a, E> CallerWrap<'a, E>
+impl<'a, E> Deref for CallerWrap<'a, E>
```

Comment:
Oh no, please no. I absolutely disagree with that.

Deref for such compound struct is not well determined but if it would have to be implemented then it would logically return `Caller` since the struct named `CallerWrap`. So two methods `memory_manager`/`memory_manager_mut` in `CallerWrap` are better then implementation of `Deref*`

Reply:
Done